builder/cpp_builder.rb

```diff
@@ -304,33 +305,58 @@ def file_to_test(filename)
   end
 
   def run_tests
-    xml_log = "#{@abs_cpp_test_out_dir}/results.xml"
+    attempt = 1
+    excluded_tests = []
+
+    loop {
+      attempt_str = @max_attempts ? "#{attempt}/#{@max_attempts}" : attempt
+
+      xml_log = "#{@abs_cpp_test_out_dir}/results-#{attempt}.xml"
+      log "run attempt #{attempt_str} (log: #{xml_log})"
+
+      # Choose test executable
+      case @mode
+      when :msbuild_windows
+        # On Windows/MSVC, executable will be nested in Debug/
+        ks_tests_bin = "#{@obj_dir}/Debug/ks_tests"
+      when :make_posix
+        # On POSIX systems, it will be directly in obj dir
+        ks_tests_bin = "#{@obj_dir}/ks_tests"
+      else
+        raise "Unknown mode=#{@mode}"
+      end
 
-    # Choose test executable
-    case @mode
-    when :msbuild_windows
-      # On Windows/MSVC, executable will be nested in Debug/
-      ks_tests_bin = "#{@obj_dir}/Debug/ks_tests"
-    when :make_posix
-      # On POSIX systems, it will be directly in obj dir
-      ks_tests_bin = "#{@obj_dir}/ks_tests"
-    else
-      raise "Unknown mode=#{@mode}"
-    end
+      tests_cli = [
+        ks_tests_bin,
+        '--log_format=XML',
+        "--log_sink=#{xml_log}",
+        '--log_level=all',
+        '--report_level=detailed',
+      ]
+      excluded_tests.each { |test| tests_cli << "--run_test=!#{test}" }
```

Comment:
This one is also a bit dangerous on Windows: I recall that command line was limited to 32KiB. If we'll concatenate all test names we have now ~7KiB worth of such strings, but it potentially plants an unpleasant surprise for the future.

The only way to work that around as far as I remember is for executable to provide alternative means to source command line args (e.g. read them from file), but boost::test does not seem to support that.

Reply:
In theory, this is a valid concern (I thought of that too, it just didn't feel that important to me) and not only on Windows, Unix-like OSes have a limit too, though probably higher than 32 KiB. In practice, I believe we are relatively far from reaching that limit (as you mentioned, even if every single test aborted, which already sounds extraordinary, we would still be well below that limit).

As you noted, I also haven't found a way to make Boost.Test read the arguments from file. But I see two ways we could solve this problem:

1. Even if Boost.Test itself doesn't support reading CLI args from file, we could probably implement this functionality ourselves by [customizing the module's entry point](https://www.boost.org/doc/libs/1_84_0/libs/test/doc/html/boost_test/adv_scenarios/shared_lib_customizations/entry_point.html). The new entry point would simply read the arguments from a file and pass them to the actual `main` function provided by Boost.Test.
2. We could take inspiration from how `xargs` works, i.e. splitting the long list of arguments into multiple batches so each batch is below the maximum limit. We cannot reliably do this with the exclusion list, but we can with an inclusion list - so we would use the [absolute specification](https://www.boost.org/doc/libs/1_84_0/libs/test/doc/html/boost_test/runtime_config/test_unit_filtering.html#boost_test.runtime_config.test_unit_filtering.absolute_specification), positively listing each test to run.

   This approach has a number of downsides compared to approach 1 - the `cpp_builder` would have to keep track of the entire set of test names, find out the command line length limit (or use `xargs` to take care of that, although that introduces a dependency) and the number of tests that we can run at once (i.e. in one batch) would be limited by how many we can list in the command line, so the test executable would have to be run multiple times just to execute a single snapshot of the test set.

Approach 1 seems much better and it should be pretty easy to implement, probably just a few lines of code.

But I don't see it as a high priority right now because it's not currently causing a problem. My view of our testing system is that it must be only as good as it proves necessary, and no better, because it's an internal tool, not meant be used by end users. We don't need to support C++ compilers / Boost versions that we don't run, or situations that have never happened or cannot happen with the current state of the project.

You know that I'm a perfectionist and I'd rather see everything working in the general case, but when it comes to the testing system, even I realize that solving all potential problems "in advance" doesn't make any difference (it's better to focus our efforts to pretty much any other KS component), only solving real problems does.